### PR TITLE
Feature: Upload File

### PR DESCRIPTION
**Pull request by YehorPytomets**
### Changes introduced by this PR:
* Implement uploading file into FileHub. Firstly file is uploaded into server, then ui responds to changes.

### CODE REVIEW

File: webclient/app/services/upload-file-service.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  selectFile() {
+    this._inputElement.click();
+  }
+
+  /**
+   * Registers on file uploaded event handler.
+   *
+   * @param {function(file: File)} handler - on-file-uploaded handler.
+   */
+  onFileUploaded(handler) {
```
Comment (dkalinin) [inline webclient/app/services/upload-file-service.js:41]: - [ ] onFileSelected

File: webclient/app/services/upload-file-service.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * Service used for uploading files into application memory.
+ */
+export default class UploadFileService {
+
+  /**
+   * Instantiates UploadFileService.
+   */
+  constructor() {
+    this._init();
```
Comment (dkalinin) [inline webclient/app/services/upload-file-service.js:12]: Please remove and call this in selectFile().

File: webclient/app/actions/upload-file-action.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+   */
+  async apply(stateManager) {
+    const {_parentFolder} = this;
+    const service = new UploadFileService();
+
+    service.onFileUploaded((file) => {
+
+      stateManager.mutate(new AddParentFolderToUploadingListMutator(_parentFolder));
+      const formData = new FormData();
+      formData.append('file', file);
```
Comment (dkalinin) [inline webclient/app/actions/upload-file-action.js:36]: - [ ] Please move to API service.

File: webclient/app/pages/file-list-explorer.js [hunk 4]
```diff
@@ -152,13 +159,15 @@ export default class FileListExplorer extends StateAwareComponent {
     this.listComponent.onUpdateListItem(async (listItem) => {
       await this.dispatch(new RenameListItemAction(listItem));
     });
+
+    this.listComponent.onUploadStart(async (folder) => await this.dispatch(new UploadFileAction(folder)));
```
Comment (dkalinin) [inline webclient/app/pages/file-list-explorer.js:163]: - [ ] async-await is not needed here.

File: webclient/app/pages/file-list-explorer.js [hunk 5]
```diff
@@ -127,12 +128,18 @@ export default class FileListExplorer extends StateAwareComponent {
...
       textContent: 'Upload File',
       title: 'Upload File',
       glyphiconClass: 'glyphicon-upload',
-      isBlocked: false,
+      isBlocked: true,
+    });
+
+    this._uploadFileButton.addClickHandler(async () => {
+      const uploadFileAction = new UploadFileAction(this.properties.stateManager.state.folder);
+      await this.dispatch(uploadFileAction);
```
Comment (dkalinin) [inline webclient/app/pages/file-list-explorer.js:142]: - [ ] async-await is not needed here.

File: webclient/app/pages/file-list-explorer.js [hunk 6]
```diff
@@ -274,6 +284,26 @@ export default class FileListExplorer extends StateAwareComponent {
         alert('folderCreatingError');
       }
     });
+
+    this.onStateChanged('parentFoldersOfUploadingFiles', (event) => {
+      const {parentFoldersOfUploadingFiles, folder, listItems} = event.detail.state;
+      _uploadFileButton.isBlocked = parentFoldersOfUploadingFiles.some((parent) => parent.id === folder.id);
+      listComponent.listItems = listItems;
```
Comment (dkalinin) [inline webclient/app/pages/file-list-explorer.js:291]: - [ ] Please remove.

File: webclient/app/pages/file-list-explorer.js [hunk 7]
```diff
@@ -152,13 +159,15 @@ export default class FileListExplorer extends StateAwareComponent {
...
   }
 
   /**
    * @inheritdoc
    */
   initState() {
-    const {listComponent, breadcrumbsComponent, _createFolderButton} = this;
+    const {listComponent, breadcrumbsComponent, _createFolderButton, _uploadFileButton} = this;
 
     this.onStateChanged('listItems', (event) => {
```
Comment (dkalinin) [inline webclient/app/pages/file-list-explorer.js:172]: - [ ] Please fix all state changed listeners to update only changed fields.

File: webclient/app/state/state.js [hunk 8]
```diff
@@ -104,4 +104,20 @@ export default class State {
...
+
+  /**
+   * Contains error, describing reason why file wasn't uploaded.
+   * @type {{notUploadedFileName: string, error: Error, parentFolder: FolderModel}}
+   */
+  uploadingFileIssue = {
+    error: null,
+    notUploadedFileName: '',
+    parentFolder: null,
+  };
```
Comment (dkalinin) [inline webclient/app/state/state.js:122]: - [ ] Default value is not needed.

File: webclient/app/services/upload-file-service.js (hunk 2 above)
Comment (YehorPytomets) [reply webclient/app/services/upload-file-service.js:12]: Done.

**Comment by YehorPytomets** [comment]
@dkalinin , PR is ready to review.

**Pull request merged**